acq4/devices/PatchPipette/states.py

```diff
@@ -1130,97 +1161,257 @@ class PatchPipetteResealState(PatchPipetteState):
+        'nuzzlePressureLimit': {'type': 'float', 'default': -1e3, 'suffix': 'Pa'},
+        'nuzzleRepetitions': {'type': 'int', 'default': 3},
+        'nuzzleSpeed': {'type': 'float', 'default': 5e-6, 'suffix': 'm/s'},
+        'pressureChangeRate': {'type': 'float', 'default': 0.5e3 / 60, 'suffix': 'Pa/s'},
+        'resealTimeout': {'type': 'float', 'default': 10 * 60, 'suffix': 's'},
+        'retractionPressure': {'type': 'float', 'default': -4e3, 'suffix': 'Pa'},
+        'retractionSpeed': {'type': 'float', 'default': 0.3e-6, 'suffix': 'm/s'},
+        'retractionSuccessDistance': {'type': 'float', 'default': 200e-6, 'suffix': 'm'},
+        'retractionSuccessState': {'type': 'str', 'default': 'home with nucleus'},
+        'secondsTestPulseAverage': {'type': 'float', 'default': 20, 'suffix': 's'},
     }
 
     def __init__(self, *args, **kwds):
-        self.retractionFuture = None
         PatchPipetteState.__init__(self, *args, **kwds)
+        self._retractionFuture = None
+        self._pressureFuture = None
+        self._updateRollingResistanceThresholds = False
+        self._rollingLongTermTestPulses = deque()
+        self._rollingRecentTestPulses = deque()
+        self._startPosition = np.array(self.dev.pipetteDevice.globalPosition())
+        self._longTermInputResistanceMean = None
+        self._longTermInputResistanceVariance = None
+        self._longTermAccessResistanceMean = None
+        self._longTermAccessResistanceVariance = None
+        self._recentInputResistanceMean = None
+        self._recentInputResistanceVariance = None
+        self._recentAccessResistanceMean = None
+        self._recentAccessResistanceVariance = None
+
+    def nuzzle(self):
+        """Wiggle the pipette around inside the cell to clear space for a nucleus to be extracted."""
+        self.setState("nuzzling")
+        if self._retractionFuture is not None:
+            self._retractionFuture.stop()
+        if self._pressureFuture is not None:
+            self._pressureFuture.stop()
+        pos = np.array(self.dev.pipetteDevice.globalPosition())
+        # TODO move back a little?
+        pipette_direction = self.dev.pipetteDevice.globalDirection()
+
+        def random_wiggle_direction():
+            """pick a random point on a circle perpendicular to the pipette axis"""
+            while np.linalg.norm(vec := np.cross(pipette_direction, np.random.uniform(-1, 1, size=3))) == 0:
+                pass  # prevent division by zero
+            return self.config['nuzzleLateralWiggleRadius'] * vec / np.linalg.norm(vec)
+
+        prev_dir = random_wiggle_direction()
+        for _ in range(self.config['nuzzleRepetitions']):
+            self.dev.pressureDevice.setPressure(source='regulator', pressure=self.config['nuzzleInitialPressure'])
+            self._pressureFuture = self.dev.pressureDevice.attainPressure(
+                target=self.config['nuzzlePressureLimit'], rate=self.config['nuzzlePressureChangeRate'])
```

Comment:
If attainPressure had a duration argument (instead of or in addition to rate), then we could drop the nuzzlePressureChangeRate parameter